src/idpi/operators/diff.py

```diff
@@ -0,0 +1,43 @@
+"""Differential operators on a regular grid."""
+
+import dataclasses as dc
+
+import numpy as np
+import xarray as xr
+
+
+@dc.dataclass
+class _Slicer:
+    ndim: int
+    axis: int
+
+    def __getitem__(self, ind):
+        slices = [slice(None)] * self.ndim
+        slices[self.axis] = ind
+        return tuple(slices)
+
+
+def _dxy(field: xr.DataArray, axis: int) -> np.ndarray:
+    s = _Slicer(ndim=len(field.dims), axis=axis)
+    df_dxy = np.empty_like(field)
+    df_dxy[s[1:-1]] = 0.5 * (field[s[2:]] - field[s[:-2]])
+    df_dxy[s[0]] = np.nan
+    df_dxy[s[-1]] = np.nan
+    return df_dxy
+
+
+def dx(field: xr.DataArray) -> xr.DataArray:
+    return field.copy(data=_dxy(field, axis=-1))
+
+
+def dy(field: xr.DataArray) -> xr.DataArray:
+    return field.copy(data=_dxy(field, axis=-2))
+
+
+def dz(field: xr.DataArray) -> xr.DataArray:
+    return field.copy(data=np.gradient(field, axis=-3))
+
+
+def dz_staggered(field: xr.DataArray) -> xr.DataArray:
```

Comment:
it would be good to add some documentation, particularly in the differences between the x,y and z and this last one that is directly changing the staggered location

Reply:
In theory, there should be staggered variants in the x and y directions too. However, due to the missing first element, I'm not sure how that boundary should be treated. The curl operator does not require the `du_dx` and `dv_dy` terms so I don't think we can rely on that specific implementation in fieldextra. I think it's wiser to implement them when we come across an actual use case.